### PR TITLE
🐛 Source Shopify: Allow the known `HTTP errors` to be retried more than once for the BULK streams

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Given this connection, we frequently see the `Internal Server Errors` or `Unknown Errors`, which could and should be retried more than once to overcome the server availability and not to raise the error in between, allowing the source to retry more than once. 

## How
- added `bulk_retry_on_exception` decorator to catch and retry the known bulk errors (max attempts = 6)
- made small refactoring to the `job_manager` to:
    - simplify the `checking the state of RUNNING job`
    - made the private methods to be `_private` when they are not used elsewhere but the module itself
    - updated the `unit_tests` to cover the changes (the logic didn't change!)
    - added a new test to cover the `retry` for known HTTP errors for BULK job: `test_job.py > test_retry_on_job_exception`

## User Impact
No impact is expected.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 1]
```diff
@@ -302,6 +312,18 @@ def on_access_denied_job(self, **kwagrs) -> AirbyteTracedException:
     def on_job_with_errors(self, errors: List[Mapping[str, Any]]) -> AirbyteTracedException:
         raise ShopifyBulkExceptions.BulkJobUnknownError(f"Could not validate the status of the BULK Job `{self.job_id}`. Errors: {errors}.")
 
+    def on_retryable_error(self, response: requests.Response, exception: Exception) -> Optional[requests.Response]:
+        if self._job_retry_on_error_count == self._job_retry_on_error_max_limit:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:316]: So my understanding is that this job retry mechanism is shared for all the slices. So if we have 10 slices and each fails only once, the stream will be incomplete, is that correct? I'm wondering if we should change that or update the naming of the variable so that it's clear that it is shared for all the slices

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 2]
```diff
@@ -335,11 +348,8 @@ def job_track_running(self) -> Union[AirbyteTracedException, requests.Response]:
             else:
                 # execute ERRORS scenario
                 self.on_job_with_errors(errors)
-        except (
-            ShopifyBulkExceptions.BulkJobBadResponse,
-            ShopifyBulkExceptions.BulkJobUnknownError,
-        ) as error:
-            return self.job_one_time_retry_error(response, error)
+        except self._retryable_errors as error:
+            return self.on_retryable_error(response, error)
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:352]: This feels odd to me. Given that `on_retryable_error` only perform the HTTP request again, we lose all the rest of the processing that `job_track_running` does so:
```
            errors = self.job_check_for_errors(response)
            if not errors:
                self.job_update_state(response)
                self.job_state_to_fn_map.get(self.job_state)(response=response)
                return response
            else:
                # execute ERRORS scenario
                self.on_job_with_errors(errors)
```

... won't be executed. Are these not important on retries? 

The same comment applies to `job_healthcheck`.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 3]
```diff
@@ -443,7 +450,8 @@ def job_check(self, created_job_response: requests.Response) -> Optional[str]:
             ShopifyBulkExceptions.BulkJobFailed,
             ShopifyBulkExceptions.BulkJobTimout,
             ShopifyBulkExceptions.BulkJobAccessDenied,
-            # this one is one-time retriable
+            # this one is retriable,mbut stil needs to be raised,
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:453]: nit: typo `retriable,mbut` -> `retriable, but `

**Review by maxi297 — COMMENTED** [review]
Can you provide more information on the comments and ping me right after so that I can re-review with the added information?

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 4]
```diff
@@ -54,6 +54,8 @@ class ShopifyBulkManager:
     # to maximize the performance for multi-connection syncs and control the bulk job size within +- 1 hours (3600 sec),
     # Ideally the source will balance on it's own rate, based on the time taken to return the data for the slice.
     job_max_elapsed_time_sec: Final[float] = 1800.0
+    # retry max limit
```
Comment (strosek) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:57]: This comment may be unnecessary, as the identifier tells what the number is for. A better comment could explain why the 6. And maybe the identifier is too long. What about something like _job_error_max_retries.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 5]
```diff
@@ -83,6 +85,14 @@ class ShopifyBulkManager:
     log_job_state_msg_count: int = field(init=False, default=0)
     # one time retryable error counter
     _one_time_error_retried: bool = field(init=False, default=False)
+    # retry counter
+    _job_retry_on_error_count: int = field(init=False, default=0)
+
+    # the set of retryable errors
```
Comment (strosek) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:91]: This and the above comments seem unecessary to me, code is self-expalanatory.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 6]
```diff
@@ -83,6 +85,14 @@ class ShopifyBulkManager:
...
+    _retryable_errors: Final[set] = (
```
Comment (strosek) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:92]: The Final annotation is not very useful in making values immutable. Maybe for constants, we could use either a function that will always return the same, or we could use immutable data structures like `frozenset`.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 7]
```diff
@@ -302,6 +312,18 @@ def on_access_denied_job(self, **kwagrs) -> AirbyteTracedException:
     def on_job_with_errors(self, errors: List[Mapping[str, Any]]) -> AirbyteTracedException:
         raise ShopifyBulkExceptions.BulkJobUnknownError(f"Could not validate the status of the BULK Job `{self.job_id}`. Errors: {errors}.")
 
+    def on_retryable_error(self, response: requests.Response, exception: Exception) -> Optional[requests.Response]:
+        if self._job_retry_on_error_count == self._job_retry_on_error_max_limit:
+            self.on_job_with_errors(self.job_check_for_errors(response))
+        else:
+            # increment the attempt
```
Comment (strosek) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:319]: This is also very clear from just reading the line of code below.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 8]
```diff
@@ -302,6 +312,18 @@ def on_access_denied_job(self, **kwagrs) -> AirbyteTracedException:
...
+            self._job_retry_on_error_count += 1
+            # retry the request
```
Comment (strosek) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:321]: Maybe using more vertical spacing would help with readability. I believe the logging entry serves as an explanation of what is about to happen. And the retry_request method is clear. I suggest replacing this comment by a empty line.

**Review by strosek — APPROVED** [review]
This codes looks OK, please address Max's concerns. I also added a few suggestions to improve readability.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:316]: We have the `stream_slices()` as a generator that works with 1 slice/time. If the slice faces the retryable Error - we go with the retry until the `_job_retry_on_error_max_limit ` is reached; otherwise, we raise the Error.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 2 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:352]: It's expected! We want to be able to retry the failed request (latest one) because we track the job state for the current cycle iteration. Then, if everything is `healthy` - we reset the state and grab the new state from the incoming response for `RUNNING` job.

If the response for the RUNNING job could be validated or has some issues getting out the JSON () from the response - we should not go with the logic you mention but retry. Then, if the retried request is OK, we go with the rest of the logic.

I'm starting to think about how we can simplify the whole BULK flow, because I find it confusing for everyone around, including me, sometimes)

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 4 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:57]: Corrected, thanks!

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 6 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:92]: I don't see the value in making set > frozen set because no one will change that in the middle of the source is running. However, I'll follow your comment and apply it - it seems reasonable from the part of best practices.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 4 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:57]: 6 - because the CDK `max_retries` default value is 6 as well.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 1 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:316]: That makes sense to me but don't we share the same job_manager for all the slices in a stream? It seems like all slices should be able to fail at least once without the stream breaking but I don't have much to confirm that. Example:

```
* Processing slice 1:
    * First call fail
    * Retry succeed
* Processing slice 2:
    * First call fail
    * Retry succeed
...
* Processing slice n:
    * First call fail
    * Retry succeed
```

In the case, since the API is still working from time to time, it seems like we should continue. I agree though that we should avoir cases like:
```
* Processing slice 1:
    * First call fail
    * Retry fail
    * Retry fail again
    <...>
```

Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:316]: That is correct, if the retried slice is succeeded - we go for the next one, parsing the result for the succeeded one before that.

Comment (maxi297) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:316]: But `_job_retry_on_error_count` is never reset to 0 between the slices, right? This would mean that the slices share the same count

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 2 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:352]: Yeah, it is pretty complex...

To go back to my question above, let's say we have:

* `response = track_running_job.request(**status_args)` fails with an error within `_retryable_errors`, we will therefore call `self.on_retryable_error(response, error)`
* We call `response = track_running_job.request(**status_args)` again and it returns `{"errors": ["an error"]}`
    * Now, this seems very dangerous because `self.job_manager.job_check(response)` will return `None` and `process_bulk_results` will silently fail by not returning records
    
Hence, I think this way of retrying does not work. We should retry the whole method and not just the HTTP request

**Review by maxi297 — COMMENTED** [review]
I think that until I understand the flow, I'll have to mark this as "Request changes". Based on what I see, there is potential for data loss which I'm not comfortable with. We can sync on this if it's easier to address

**Review by maxi297 — CHANGES_REQUESTED** [review]
Forgot to add the flag based on https://github.com/airbytehq/airbyte/pull/37589#pullrequestreview-2025500772

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 2 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:352]: > We call response = track_running_job.request(**status_args) again and it returns {"errors": ["an error"]}
Now, this seems very dangerous because self.job_manager.job_check(response) will return None and process_bulk_results will silently fail by not returning records


Based on the logic, we will `raise ShopifyBulkExceptions.BulkJobUnknownError()` with the details of what are these errors, like this:
```python
raise ShopifyBulkExceptions.BulkJobUnknownError(f"Could not validate the status of the BULK Job `123`. Errors: an error.")
```
In this scenario, the source will be stopped, we don't case about the response here, because we cannot move forward with the Errors, The platform will trigger the new attempt. When such things happen for Shopify - we should not just swallow that error but raise it. Didn't find anything dangerous retrying the latest request. 

It's clear if the Server returns 200 responses with the details about the actual Errors for the Bulk Job not being able to succeed. This is not user-actionable. Thus the error should be raised. We should check that and find the approach to fix it.

**Comment by bazarnov** [comment]
Would be happy to discuss this once you're available.

> I think that until I understand the flow, I'll have to mark this as "Request changes". Based on what I see, there is potential for data loss which I'm not comfortable with. We can sync on this if it's easier to address

Thank you for the review.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py]: Oh, got it, finally! 
Added the reset for that counter when we successfully got out of the loop with the successful request after retry.

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 2 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:351]: I still think it does not work as expected. To leave the while loop in `job_check_state`,  we need to `self.job_completed() -> True`. In order to do that, `self.job_state_to_fn_map.get(self.job_state)(response=response)` needs to be called. It will not be called in the case of a retry as it is only called [here](https://github.com/airbytehq/airbyte/pull/37589/files#diff-69c119a9dc0a2f8a93bc200b41dfe1343d3ac94f117698a9442f30c5e2cadd60L330-L333). Can you add a test to validate this understanding?

Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py]: Changed the logic of checking the Bulk Job State, please check the code once again. Thank you.

File: airbyte-integrations/connectors/source-shopify/unit_tests/graphql_bulk/test_job.py [hunk 9]
```diff
@@ -167,38 +167,43 @@ def test_job_check(mocker, request, requests_mock, job_response, auth_config, er
...
-    # get job_id from FIXTURE
-    job_id = request.getfixturevalue(job_response).get("data", {}).get("node", {}).get("id")
     # patching the method to get the right ID checks
-    if job_id:
-        mocker.patch("source_shopify.shopify_graphql.bulk.job.ShopifyBulkManager.job_get_id", value=job_id)
+    stream.job_manager._job_id = request.getfixturevalue(job_response).get("data", {}).get("node", {}).get("id")
+    stream.job_manager._job_backoff_time = 0
+    stream.job_manager._job_max_retries = 2
+    # making the check method to raise the specific error
+    mocker.patch(
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/unit_tests/graphql_bulk/test_job.py:188]: `_job_check_for_errors` seems like an implementation detail. Why don't we mock the HTTP request instead?

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py [hunk 10]
```diff
@@ -111,247 +109,209 @@ def job_size_adjusted_expand_factor(self, coef: float = 0.5) -> float:
...
-        except (
-            ShopifyBulkExceptions.BulkJobBadResponse,
-            ShopifyBulkExceptions.BulkJobUnknownError,
-        ) as error:
-            return self.job_one_time_retry_error(response, error)
+    def _job_send_state_request(self) -> requests.Response:
+        with self.session as job_state_request:
+            status_args = self._job_get_request_args(ShopifyBulkTemplates.status)
+            self._request = requests.Request(**status_args, auth=self.session.auth).prepare()
+            return job_state_request.send(self._request)
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:302]: Should there be a try/except here in case of ConnectionError?

File: airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py [hunk 11]
```diff
@@ -642,7 +642,7 @@ def __init__(self, config: Dict) -> None:
         # define BULK Manager instance
         self.job_manager: ShopifyBulkManager = ShopifyBulkManager(
             session=self._session,
-            base_url=f"{self.url_base}/{self.path()}",
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py:645]: Why has this changed? It does not seem related to the retrying part

**Review by maxi297 — COMMENTED** [review]
I prefer the newest changes. I like that:
* method visibility (private or public) is now explicit
* the retry is with an annotation on the `_job_check_state` method

I have a couple of concerns which you can push back on but once we see the same path forward, I'll approve

File: airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py (hunk 11 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py:645]: It was like that:

before: `http://domain.my_shopify.com//graphql.json` - introducing the extra `/` in the path. It seems like the `self.path()` includes this once joined with the `base_url`. No need for that.

After: `http://domain.my_shopify.com/graphql.json`

File: airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py (hunk 10 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:302]: We can probably wait until we have the CDK level of handler for the `ConnectionError`, and that would be covered out-of-the-box.

File: airbyte-integrations/connectors/source-shopify/unit_tests/graphql_bulk/test_job.py (hunk 9 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/unit_tests/graphql_bulk/test_job.py:188]: updated.

**Review by maxi297 — APPROVED** [review]
Cool, thanks for the changes. :shipit: !
